File: src/TypeFormatter/LiteralUnionTypeFormatter.ts
import type { Definition } from "../Schema/Definition.js";
import type { RawTypeName } from "../Schema/RawType.js";
import type { SubTypeFormatter } from "../SubTypeFormatter.js";
import type { BaseType } from "../Type/BaseType.js";
import { EnumType } from "../Type/EnumType.js";
import { LiteralType, type LiteralValue } from "../Type/LiteralType.js";
import { NullType } from "../Type/NullType.js";
import { StringType } from "../Type/StringType.js";
import { UnionType } from "../Type/UnionType.js";
import { typeName } from "../Utils/typeName.js";
import { toEnumType } from "./EnumTypeFormatter.js";

export class LiteralUnionTypeFormatter implements SubTypeFormatter {
    public supportsType(type: BaseType): boolean {
        return type instanceof UnionType && type.getTypes().length > 0 && isLiteralUnion(type);
    }

    public getDefinition(unionType: UnionType): Definition {
        let hasString = false;
        let preserveLiterals = false;
        let allStrings = true;
        let hasNull = false;

        const literals = unionType.getFlattenedTypes();

        // filter out String types since we need to be more careful about them
        const types = literals.filter((literal) => {
            if (literal instanceof StringType) {
                hasString = true;
                preserveLiterals ||= literal.getPreserveLiterals();
                return false;
            }

            if (literal instanceof NullType) {
                hasNull = true;
                return true;
            }

            if (literal instanceof LiteralType && !literal.isString()) {
                allStrings = false;
            }

            return true;
        });

        if (allStrings && hasString && !preserveLiterals) {
            return hasNull ? { type: ["string", "null"] } : { type: "string" };
        }

        const typeValues: Set<LiteralValue | null> = new Set();
        const typeNames: Set<RawTypeName> = new Set();

        for (const type of types) {
            appendTypeNames(type, typeNames);
            appendTypeValues(type, typeValues);
        }

        const schema = {
            type: toEnumType(Array.from(typeNames)),
            enum: Array.from(typeValues),
        };

        return preserveLiterals ? { anyOf: [{ type: "string" }, schema] } : schema;
    }

    public getChildren(): BaseType[] {
        return [];
    }
}

export function isLiteralUnion(type: UnionType): boolean {
    return type
        .getFlattenedTypes()
        .every(
            (item) =>
                item instanceof LiteralType ||
                item instanceof NullType ||
                item instanceof StringType ||
                item instanceof EnumType,
        );
}

/**
 * Appends all possible type names of a type to the given set.
 */
function appendTypeNames(type: BaseType, names: Set<RawTypeName>) {
    if (type instanceof EnumType) {
        for (const value of type.getValues()) {
            names.add(typeName(value));
        }

        return;
    }

    if (type instanceof LiteralType) {
        names.add(typeName(type.getValue()));
        return;
    }

    names.add(typeName(null));
}

/**
 * Appends all possible values of a type to the given set.
 */
function appendTypeValues(type: BaseType, values: Set<LiteralValue | null>) {
    if (type instanceof EnumType) {
        for (const value of type.getValues()) {
            values.add(value);
        }

        return;
    }

    if (type instanceof LiteralType) {
        values.add(type.getValue());
        return;
    }

    values.add(null);
}

File: internal/formatter/literal_union_type_formatter.go
package formatter

import (
	"github.com/vega/ts-json-schema-generator/internal/schema"
	"github.com/vega/ts-json-schema-generator/internal/types"
)

// LiteralUnionTypeFormatter mirrors src/TypeFormatter/LiteralUnionTypeFormatter.ts.
type LiteralUnionTypeFormatter struct{}

func NewLiteralUnionTypeFormatter() *LiteralUnionTypeFormatter {
	return &LiteralUnionTypeFormatter{}
}

func (f *LiteralUnionTypeFormatter) SupportsType(t types.Type) bool {
	unionType, ok := t.(*types.UnionType)
	return ok && len(unionType.Types()) > 0 && types.IsLiteralUnion(unionType)
}

func (f *LiteralUnionTypeFormatter) GetDefinition(t types.Type) *schema.Definition {
	unionType := t.(*types.UnionType)

	hasString := false
	preserveLiterals := false
	allStrings := true
	hasNull := false

	literals := unionType.FlattenedTypes(nil)

	// Filter out string types since we need to be more careful about them.
	filtered := make([]types.Type, 0, len(literals))
	for _, literal := range literals {
		switch item := literal.(type) {
		case *types.StringType:
			hasString = true
			preserveLiterals = preserveLiterals || item.PreserveLiterals
			continue
		case *types.NullType:
			hasNull = true
		case *types.LiteralType:
			if !item.IsString() {
				allStrings = false
			}
		}
		filtered = append(filtered, literal)
	}

	if allStrings && hasString && !preserveLiterals {
		if hasNull {
			return &schema.Definition{Type: []string{"string", "null"}}
		}
		return &schema.Definition{Type: "string"}
	}

	typeValues := []any{}
	seenValues := map[any]bool{}
	typeNames := []string{}
	seenNames := map[string]bool{}

	appendName := func(name string) {
		if !seenNames[name] {
			seenNames[name] = true
			typeNames = append(typeNames, name)
		}
	}
	appendValue := func(value any) {
		if !seenValues[value] {
			seenValues[value] = true
			typeValues = append(typeValues, value)
		}
	}

	for _, item := range filtered {
		switch member := item.(type) {
		case *types.EnumType:
			for _, value := range member.Values {
				appendName(typeName(value))
			}
		case *types.LiteralType:
			appendName(typeName(member.Value))
		default:
			appendName("null")
		}
		switch member := item.(type) {
		case *types.EnumType:
			for _, value := range member.Values {
				appendValue(value)
			}
		case *types.LiteralType:
			appendValue(member.Value)
		default:
			appendValue(nil)
		}
	}

	var definition *schema.Definition
	if len(typeNames) == 1 && len(typeValues) == 1 {
		definition = &schema.Definition{Type: toEnumType(typeNames), Const: schema.Ptr(typeValues[0])}
	} else {
		definition = &schema.Definition{Type: toEnumType(typeNames), Enum: typeValues}
	}

	if hasString {
		return &schema.Definition{AnyOf: []*schema.Definition{{Type: "string"}, definition}}
	}
	return definition
}

func (f *LiteralUnionTypeFormatter) GetChildren(t types.Type) []types.Type { return nil }
